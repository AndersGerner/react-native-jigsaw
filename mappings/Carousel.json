[{"name":"Carousel","tag":"Carousel","category":"blocks","description":"A horizontal scrolling carousel of images","preview_image_url":"https://res.cloudinary.com/altos/image/upload/draftbit/library/jigsaw-1.0/reps/Carousel.png","supports_list_render":false,"layout":{"height":250},"props":{"images":{"label":"Images","description":"A series of images to display in the carousel","editable":true,"type":"localImageArray","value":null},"resizeMode":{"label":"Resize Mode","description":"Determines how to resize the images when the frame doesn't match the raw image dimensions","editable":true,"required":false,"value":"cover","type":"flatArray","options":["cover","contain","stretch","repeat","center"]},"dotColor":{"label":"Dot Color","description":"Color of the carousel's dots","editable":true,"required":true,"value":"strong","type":"color"}}}]