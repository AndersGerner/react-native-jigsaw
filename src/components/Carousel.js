import React from "react";
import { ScrollView, Image, View, StyleSheet } from "react-native";
import Elevation from "./Elevation";
import { withTheme } from "../core/theming";
import { COMPONENT_TYPES, FORM_TYPES } from "../core/component-types";
import type { Theme } from "../types";
import Config from "./Config";

export type CarouselProps = {
  images: Array<string | { uri: string }>,
  swiperPalette: "surface" | "background",
  aspectRatio: number,
  theme: Theme,
  style: any
};

class Carousel extends React.PureComponent<CarouselProps> {
  state = {};

  static defaultProps = {
    images: [
      Config.cardImageUrl,
      Config.cardImageUrl,
      Config.cardImageUrl,
      Config.cardImageUrl,
      Config.cardImageUrl,
      Config.cardImageUrl,
      Config.cardImageUrl
    ],
    aspectRatio: 1.25,
    swiperPalette: "background"
  };

  constructor(props) {
    super(props);

    this.state = {
      scrollOffset: 0
    };
  }

  handleScroll = e => {
    this.setState({ scrollOffset: e.nativeEvent.contentOffset.x });
  };

  onPageLayout = event => {
    const { width } = event.nativeEvent.layout;
    this.setState({ width });
  };

  render() {
    const { images, aspectRatio, swiperPalette, theme, style } = this.props;
    const { colors, spacing } = theme;
    const { width } = this.state;

    return (
      <View
        style={[styles.container, style, { aspectRatio }]}
        onLayout={this.onPageLayout}
      >
        <ScrollView
          onScroll={this.handleScroll}
          horizontal
          pagingEnabled
          showsHorizontalScrollIndicator={false}
        >
          {images.map((image, index) => (
            <View style={[styles.slidingPanel, { width }]} key={index}>
              <Image
                source={typeof image === "string" ? { uri: image } : image}
                resizeMode="cover"
                style={{ width, aspectRatio }}
              />
            </View>
          ))}
        </ScrollView>
        <View style={[styles.swipeNavWrapper, { bottom: spacing.large }]}>
          <View style={styles.swipeNav}>
            {[...Array(images.length)].map((_, i) => {
              const calculatedIndex = this.state.scrollOffset / width;
              const activeDot =
                calculatedIndex >= i - 0.5 && calculatedIndex < i + 0.5;

              let backgroundColor;
              if (swiperPalette === "surface") {
                if (activeDot) {
                  backgroundColor = colors.strong;
                } else {
                  backgroundColor = colors.light;
                }
              } else {
                if (activeDot) {
                  backgroundColor = colors.background;
                } else {
                  backgroundColor = colors.surface;
                }
              }

              return (
                <Elevation
                  key={i}
                  style={[
                    { backgroundColor, marginHorizontal: spacing.small / 2 },
                    styles.dot,
                    activeDot ? styles.activeDot : null
                  ]}
                />
              );
            })}
          </View>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    alignSelf: "stretch"
  },
  slidingPanel: {
    alignItems: "center",
    justifyContent: "center"
  },
  swipeNavWrapper: {
    position: "absolute",
    justifyContent: "center",
    alignItems: "center",
    left: 0,
    right: 0
  },
  swipeNav: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "center"
  },
  dot: {
    width: Config.swiperInactiveDotSize,
    height: Config.swiperInactiveDotSize,
    borderRadius: Config.swiperInactiveDotSize
  },
  activeDot: {
    opacity: 1,
    elevation: 1,
    width: Config.swiperActiveDotSize,
    height: Config.swiperActiveDotSize,
    borderRadius: Config.swiperActiveDotSize
  }
});

export default withTheme(Carousel);

export const SEED_DATA = [
  {
    name: "Carousel",
    tag: "Carousel",
    category: COMPONENT_TYPES.carousel,
    description: "A horizontal scrolling carousel of images",
    preview_image_url: "{CLOUDINARY_URL}/Carousel.png",
    supports_list_render: false,
    layout: {
      width: "100%",
      height: 250
    },
    props: {
      images: {
        label: "Images",
        description: "A series of images to display in the carousel",
        editable: true,
        type: FORM_TYPES.localImageArray,
        value: null
      },
      swiperPalette: {
        label: "Swiper palette",
        description: "The palette type to use for the swiper",
        editable: true,
        type: FORM_TYPES.flatArray,
        value: "background",
        options: ["background", "surface"],
        required: true
      },
      aspectRatio: {
        label: "Aspect ratio",
        description: "Aspect ratio of the carousel",
        type: FORM_TYPES.aspectRatio,
        value: 1.25,
        editable: true,
        required: true
      }
    }
  }
];
