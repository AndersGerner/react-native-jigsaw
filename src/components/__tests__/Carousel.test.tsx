import * as React from "react";
import renderer from "react-test-renderer";
import Carousel from "../Carousel";

it("renders carousel", () => {
  const tree = renderer
    .create(
      <Carousel
        images={[
          require("../../assets/bg.png"),
          require("../../assets/bg.png"),
        ]}
      />
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
