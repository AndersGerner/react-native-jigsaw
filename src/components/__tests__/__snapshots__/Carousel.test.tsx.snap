// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders carousel 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignSelf": "stretch",
      },
      Object {
        "aspectRatio": undefined,
        "width": 0,
      },
      Object {
        "height": 375,
      },
    ]
  }
>
  <RCTScrollView
    horizontal={true}
    onScroll={[Function]}
    pagingEnabled={true}
    scrollEventThrottle={16}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "width": 0,
            },
          ]
        }
      >
        <Image
          resizeMode="cover"
          source={1}
          style={
            Object {
              "aspectRatio": undefined,
              "flex": 1,
              "height": 375,
              "width": 0,
            }
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "width": 0,
            },
          ]
        }
      >
        <Image
          resizeMode="cover"
          source={1}
          style={
            Object {
              "aspectRatio": undefined,
              "flex": 1,
              "height": 375,
              "width": 0,
            }
          }
        />
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "left": 0,
          "position": "absolute",
          "right": 0,
        },
        Object {
          "bottom": 16,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(255, 255, 255, 1)",
            "borderRadius": 6,
            "height": 6,
            "marginHorizontal": 4,
            "opacity": 0.5,
            "width": 6,
            "zIndex": 100,
          }
        }
      />
      <View
        style={
          Object {
            "backgroundColor": "rgba(255, 255, 255, 1)",
            "borderRadius": 6,
            "height": 6,
            "marginHorizontal": 4,
            "opacity": 0.5,
            "width": 6,
            "zIndex": 100,
          }
        }
      />
    </View>
  </View>
</View>
`;
